frontend/packages/topology/src/components/quick-search/QuickSearchModalBody.tsx

```diff
@@ -0,0 +1,150 @@
+    getQueryArgument('catalogSearch') || '',
+  );
+  const [selectedItemId, setSelectedItemId] = React.useState<string>('');
+  const [selectedItem, setSelectedItem] = React.useState<CatalogItem>(null);
+  const listCatalogItems = catalogItems?.slice(0, 5);
+  const ref = React.useRef<HTMLDivElement>(null);
+
+  React.useEffect(() => {
+    searchTerm && allCatalogItemsLoaded && setCatalogItems(searchCatalog(searchTerm));
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [allCatalogItemsLoaded, searchCatalog]);
+
+  React.useEffect(() => {
+    if (catalogItems && !selectedItemId) {
+      setSelectedItemId(catalogItems?.[0]?.uid);
+      setSelectedItem(catalogItems?.[0]);
+    }
+  }, [catalogItems, selectedItemId]);
+
+  const onSearch = React.useCallback(
+    (value: string) => {
+      setSearchTerm(value);
+      setCatalogItems(value ? searchCatalog(value) : null);
+      setSelectedItemId('');
+      setSelectedItem(null);
+    },
+    [searchCatalog],
+  );
+
+  const onEscapePress = React.useCallback(() => {
+    const searchInput = ref?.current?.firstElementChild?.children?.[1] as HTMLInputElement;
+    if (searchInput?.value) {
+      document.activeElement !== searchInput && searchInput.focus();
+      onSearch('');
+    } else {
+      closeModal();
+      removeQueryArgument('catalogSearch');
+    }
+  }, [closeModal, onSearch]);
+
+  const getIndexOfSelectedItem = React.useCallback(
+    () => listCatalogItems?.findIndex((item) => item.uid === selectedItemId),
+    [listCatalogItems, selectedItemId],
+  );
+
+  const selectPrevious = React.useCallback(() => {
+    const index = getIndexOfSelectedItem();
+    setSelectedItemId(listCatalogItems?.[index - 1]?.uid);
+    setSelectedItem(listCatalogItems?.[index - 1]);
+  }, [listCatalogItems, getIndexOfSelectedItem]);
+
+  const selectNext = React.useCallback(() => {
+    const index = getIndexOfSelectedItem();
+    setSelectedItemId(listCatalogItems?.[index + 1]?.uid);
+    setSelectedItem(listCatalogItems?.[index + 1]);
+  }, [listCatalogItems, getIndexOfSelectedItem]);
+
+  React.useEffect(() => {
+    const onKeyDown = (e: KeyboardEvent) => {
+      switch (e.code) {
+        case 'Escape': {
+          onEscapePress();
+          break;
+        }
+        case 'ArrowUp': {
+          selectPrevious();
+          break;
+        }
+        case 'ArrowDown': {
+          selectNext();
+          break;
+        }
+        default:
+      }
+    };
+
+    const onOutsideClick = (e: MouseEvent) => {
+      const searchInput = ref?.current?.firstElementChild?.children?.[1] as HTMLInputElement;
+      if (!ref?.current?.contains(e.target as Node)) {
+        if (searchInput?.value) {
+          setQueryArgument('catalogSearch', searchInput.value);
+        } else {
+          removeQueryArgument('catalogSearch');
+        }
+        closeModal();
+      }
+    };
+
+    document.addEventListener('click', onOutsideClick);
```

Comment:
Interesting... Odd the Modal from PF doesn't _just_ support this 🤔 

I wonder if there should be some separation of concerns here... this component is for the content of the modal but is doing some "modal as a whole" actions (like closing via outside/mask click & keyboard controls). 

I'm thinking:

- `QuickStartModal` handles mask/outside clicks
- `QuickStartModal` handles escape keyboard press
- `QuickStartModalBody` handles up/down for selection of items

Thoughts?

Reply:
Yeah initially I was going to structure the event handlers this way but if you see the events on Modal and ModalBody are not really separated from each other. 

Like when the user clicks outside we are supposed to close the modal and if there is a value in the search bar we have to save it and present it to the user next time they open the modal. So one action happens on the Modal and the other on ModalBody. 

Similarly in case of escape key press we are supposed to clear the text if there is any text in the search bar and if then the user again presses escape we have to close the modal.

Handling these events in the body itself made more sense to me as the search bar state is being handled in the body and can be easily updated accordingly. But I am open to suggestions if you have any.

Reply:
🤔  Valid thoughts... I didn't follow it all the way through and you did haha... Perhaps we leave it as is for now. Sounds like there is an avenue to doing this, but it may be a bit messy. I'm more sad PF doesn't support clicking on a mask to close the modal...

Reply:
Yeah definitely it can be done in the modal itself but that would make the code really complex as we will have to gain reference for the search bar in the modal and then use/edit its value from there.